### PR TITLE
perf: add support for perfetto performance tracing

### LINKED ISSUE #16988: benchmarking: add ability to trace Envoy code paths
benchmarking - add ability to trace Envoy code paths

*Description*:
It would be useful to be able to instrument Envoy code with tracing elements that time code execution in order to measure time spent in any given code path. The desired output would look similar to a flamegraph, but specific to codepaths that are instrumented.

This would allow developers to be more targeted in their benchmarking, and would enable timing deployments that don't have access to flamegraph or similar tools (e.g., inaccessible remote servers, mobile deployments via envoy mobile).

*Potential Implementation*:
Data Collection: data collection would happen with a RAII class similar to Scope Trackers that can create recursively nested scopes that time execution from creation to destruction. Further, implementing timing as a scope stack lends nicely to constructing rich timelines of code execution. This class could use the existing Timespan interface to produce timing measurements.

Data retrieval: a new extension similar to metrics sinks, or access loggers could be built in order to retrieve timing information.

*Relevant Links*
The desired instrumentation would be similar to that in https://github.com/tokio-rs/tracing. More research is required to see if there is an already existing similar library in C++, or if a simple custom solution as described above would suffice.

### PR DESCRIPTION
**Pull request by rojkov**
Commit Message: perf: add support for perfetto performance tracing
Additional Description:
Add a new build option enabling support for Perfetto SDK which can be used to trace Envoy code paths.
Risk Level: low
Testing: manual
Docs Changes: add notes to bazel/PPROF.md
Release Notes: updated current.rst
Platform Specific Features: n/a
Fixes #16988

### CODE REVIEW

**Comment by ggreenway** [comment]
@rojkov are you ok if this PR is stalled until @junr03 is back, waiting on an answer to your question?

/wait-any

**Comment by rojkov** [comment]
@ggreenway Yes, I'm fine to wait.

**Comment by rojkov** [comment]
@junr03 PTAL

**Comment by rojkov** [comment]
@junr03 Yeap, I don't expect any problems when it's disabled. My concern was that perfetto fires up a thread upon `Envoy::Server::InistanceImpl::initialize()` and I wasn't sure Envoy Mobile triggers this code path.

**Comment by junr03** [comment]
@rojkov Envoy Mobile does (indirectly) use `Envoy::Server::InstanceImpl::initialize()`. But as I see it all the perfetto code is guarded [there](https://github.com/envoyproxy/envoy/pull/17591/files#diff-a39ff317d26d008f20bca53f5e3f4e171d169a5702f82409ca941bddaac46064R400), right?

**Comment by rojkov** [comment]
> But as I see it all the perfetto code is guarded there, right?

Yes, it is. Everything is guarded build time.

**Review by junr03 — APPROVED** [review]

**Comment by rojkov** [comment]
@moderation is there something I could improve for the new dep?

**Comment by moderation** [comment]
/lgtm deps

As this is guarded at build time with `bazel --define=perf_tracing=enabled` it looks OK from a deps perspective. Would be good for @htuch review

File: source/server/server.cc
Comment (htuch) [inline source/server/server.cc:409]: Since tracers only seem to be written at server destruction, I'm a bit confused:
1. What is the meaning of this buffer vs. presumably the indefinite trace accumulation in memory?
2. What if a user wants to capture snapshots of performance traces for a live server?

**Review by htuch — COMMENTED** [review]
Looks good, just a few nits/comments.

**Comment by htuch** [comment]
Also, how does this PR relate to https://github.com/envoyproxy/envoy/blob/main/source/common/common/perf_annotation.h? CC @jmarantz 

Maybe there is an opportunity to reuse those annotations or migrate them to the Perfetto ones?

**Comment by rojkov** [comment]
Thanks! I'll address the comments next week.
/wait

File: source/server/server.cc
Comment (rojkov) [reply source/server/server.cc]: I have added a new string field to Bootstrap, but maybe the field could be a message including the file path and Perfetto's `TraceConfig` message. But not all `TraceConfig`'s fields are relevant to Envoy (right now at least).

File: source/common/common/perf_tracing.h
Comment (jmarantz) [reply source/common/common/perf_tracing.h:44]: Per earlier comment I think it'd be good to understand when to use the existing source/common/common/perf_annotation.h and when to use this, or, assuming this is a lot better, remove perf_annotation.h and replace calls to it existing in the codebase with calls to this.

File: source/server/server.cc
Comment (rojkov) [reply source/server/server.cc:400]: By default it's a ring buffer that needs to be flushed to a file (or another sink) before it gets overwritten. I've modified the behavior to write to the file continuously in the tracing thread. Now the file gets just flushed and closed at server destruction.

Currently switching tracing on and off for a live server is not implemented as I use this SDK for development only. But this functionality could be added through the admin interface.

File: bazel/repository_locations.bzl
Comment (rojkov) [reply bazel/repository_locations.bzl:126]: done

File: source/common/common/perf_tracing.h
Comment (rojkov) [reply source/common/common/perf_tracing.h:44]: I added a couple of macros to `ActiveStream` - one scoped macro and a pair of macros to trace an object's life span - as an illustration.

I think the purposes of perf annotations and this perf tracing are different a bit. The former works great for benchmarking, the latter gives better insight on what happens in what order and how much time it takes. Also its cross-scoped macros don't work well for open-loop benchmarks (`TRACE_EVENT_BEGIN` for a single category can't be nested in a single track). I'd keep both.

Comment (jmarantz) [reply source/common/common/perf_tracing.h:44]: OK, that's fine. Can we, in this PR, put comments in both headers to xref one another and describe when you'd use one vs the other? Just from the naming of them it's hard to discern when to use each one.

Comment (htuch) [reply source/common/common/perf_tracing.h:50]: I guess I'm not fully convinced yet that perf annotations can't be replaced with tracing. @rojkov can you explain the open loop benchmark situation further?

Comment (rojkov) [reply source/common/common/perf_tracing.h:50]: In open-loop benchmarking new requests are fired regardless of completion of any requests sent previously. For example two requests are sent one after another, but the second one is sent before the first response is received.

Now consider putting `TRACE_EVENT_BEGIN("core", "ActiveStreamLifeSpan")` into `ActiveStream`'s constructor and `TRACE_EVENT_END("core")` into its destructor. The first request will emit one opening event for "ActiveStreamLifeSpan". The second request will emit one more opening event. Now there are two opening events in the stack. The response for the first request will emit the first closing event. Here's the problem: it will close the most recent opening event in the stack which corresponds to the second request. And BTW it will close the most recent opening event irrespective of its name. To avoid that the corresponding opening and closing events should be placed on separate tracks in addition to the existing thread tracks. In this particular case every new request creates at least one new track. Benchmarks may create hundreds or thousands of such tracks. I use https://ui.perfetto.dev/ for analysis and it simply doesn't show anything for trace files with so many tracks. And if it did such a visualization would be too cluttered and useless.

Indeed this issue could be circumvented with handcrafted SQL queries, I guess, or even with custom trace-based metrics added to the Trace Processor, but perf annotations "just work" in this case.

I'm going to learn more about extending the Trace Processor with custom metrics, but until then dropping perf annotations would be premature I guess.

Comment (htuch) [reply source/common/common/perf_tracing.h:50]: Thanks for the detailed explanation. I agree, this seems broken and the basic issue here is that (correct me if I'm wrong) we're an event-based system and the trace framework isn't tracking context across events on the same thread. If we try and teach it this, it produces a large number of independent traces that the tooling doesn't do a good job at synthesizing. Is this the gist?

If so, I'm wondering how this is ever useful in an Envoy worker thread. The problem doesn't seem confined to open-loop benchmarking. What's a situation in which the cross-request tracking doesn't act in a misleading way?

Comment (rojkov) [reply source/common/common/perf_tracing.h:50]: You understood it correctly.

Perfetto produces easily consumable data for cross-stack events in case of consecutive requests produced by one virtual user (`--concurrency=1` in terms of nighthawk). If we teach Perfetto to put such events on separate tracks the produced data would not be misleading still, but it would require additional manipulations to extract useful bits. This is my current understanding.

Perf annotations do a similar job for cross-stack perf operations: every such perf operation is associated with a class instance which owns it. Perfetto events can be associated with `this` too via a track specially minted for `this`. The difference is perf annotations ignore these owner-operation relationships in the final report automatically. Perfetto keeps this info requiring additional post-processing of the tracing data.

If we forget about cross-stack operations/events and use RAII-based ones only then Perfetto provides better visualization out of the box.

Also for envoy-mobile purposes the tracing info generated by Perfetto seems to be more useful.

File: bazel/PPROF.md
Comment (htuch) [inline bazel/PPROF.md:376]: Will this give a unique track per request? Or just a child track for HTTP requests in general?

File: source/common/common/perf_tracing.h
Comment (htuch) [reply source/common/common/perf_tracing.h:50]: OK, this is great. Can you somehow capture this in a form that we can make referencable the next time this discussion comes up? Maybe a dedicated issue or add to the usage docs (I know there is already track mention there, but it's hard to infer the relationship between different tools from that and when to use one vs. the other).

File: source/common/http/conn_manager_impl.cc
Comment (htuch) [inline source/common/http/conn_manager_impl.cc:706]: Would it be better to stick to RAII examples if this can't be usefully consumed by visualization tooling?

File: source/server/server.cc
Comment (htuch) [inline source/server/server.cc:413]: Can you add some line-by-line comments explaining what is going on here? "start blocking" sounds scary.

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM modulo remaining comments.

File: source/server/server.cc
Comment (rojkov) [reply source/server/server.cc:419]: Added more comments.

File: source/common/http/conn_manager_impl.cc
Comment (rojkov) [reply source/common/http/conn_manager_impl.cc]: Right. I dropped the cross-function example.

File: bazel/PPROF.md
Comment (rojkov) [reply bazel/PPROF.md]: No, not unique track per request. I agree it'd be safer to create a track per request ID. Updated.

**Review by htuch — APPROVED** [review]
Awesome, thanks so much for your patience working through this. Great addition to our perf tooling!

**Pull request merged**
